File: src/main/java/dev/vrba/slack/tenor/bot/service/slack/SlackModalService.java
package dev.vrba.slack.tenor.bot.service.slack;

import dev.vrba.slack.tenor.bot.service.slack.dto.*;
import io.micronaut.core.annotation.NonNull;
import jakarta.inject.Singleton;
import lombok.AllArgsConstructor;
import reactor.core.publisher.Mono;

import java.util.ArrayList;
import java.util.List;

@Singleton
@AllArgsConstructor
public class SlackModalService {
    @NonNull
    private final SlackApiClient client;

    @NonNull
    public Mono<SlackViewResponse> openLoadingView(
        @NonNull final String trigger,
        @NonNull final String response
    ) {
        final var request = new SlackViewOpenRequest(
            trigger,
            new SlackView(
                null,
                "modal",
                "tenor-gif-loading",
                new SlackTextBlock("Loading Tenor GIFs...", false),
                new ArrayList<>(),
                response,
                null
            )
        );

        return client.openView(request);
    }

    @NonNull
    public Mono<Void> updateSelectionView(
        @NonNull final String viewId,
        @NonNull final String imageUrl,
        @NonNull final String responseUrl
    ) {
        final var request = new SlackViewUpdateRequest(
            viewId,
            new SlackView(
                null,
                "modal",
                "tenor-gif-selection",
                new SlackTextBlock("Tenor GIF search", false),
                List.of(
                    new SlackActionsBlock(
                        List.of(
                            new SlackButtonElement(
                                new SlackTextBlock("Previous", false),
                                "previous",
                                "previous"
                            ),
                            new SlackButtonElement(
                                new SlackTextBlock("Next", false),
                                "next",
                                "next"
                            )
                        )
                    ),
                    new SlackImageBlock(
                        imageUrl,
                        "GIF"
                    )
                ),
                responseUrl,
                new SlackTextBlock("Send this GIF to Slack", false)

            )
        );

        return client.updateView(request).then();
    }
}

File: src/main/kotlin/dev/vrba/slack/service/slack/SlackModalService.kt
package dev.vrba.slack.service.slack

import dev.vrba.slack.service.slack.dto.SlackActionsBlock
import dev.vrba.slack.service.slack.dto.SlackButtonBlock
import dev.vrba.slack.service.slack.dto.SlackImageBlock
import dev.vrba.slack.service.slack.dto.SlackTextBlock
import dev.vrba.slack.service.slack.dto.SlackView
import dev.vrba.slack.service.slack.dto.SlackViewOpenRequest
import dev.vrba.slack.service.slack.dto.SlackViewResponse
import dev.vrba.slack.service.slack.dto.SlackViewUpdateRequest
import jakarta.inject.Singleton

@Singleton
class SlackModalService(
    private val client: SlackApiClient,
) {
    suspend fun openLoadingView(
        trigger: String,
        response: String,
    ): SlackViewResponse {
        val request =
            SlackViewOpenRequest(
                trigger = trigger,
                view =
                    SlackView(
                        type = "modal",
                        callback = "tenor-gif-loading",
                        title = SlackTextBlock(text = "Loading GIFs..."),
                        metadata = response,
                    ),
            )

        return client.openView(request)
    }

    suspend fun updateSelectionView(
        viewId: String,
        imageUrl: String,
        responseUrl: String,
    ) {
        val request =
            SlackViewUpdateRequest(
                id = viewId,
                view =
                    SlackView(
                        type = "modal",
                        callback = "tenor-gif-selection",
                        title = SlackTextBlock(text = "Tenor GIF search"),
                        submit = SlackTextBlock(text = "Send this GIF to Slack"),
                        blocks =
                            listOf(
                                SlackActionsBlock(
                                    elements =
                                        listOf(
                                            SlackButtonBlock(
                                                text = SlackTextBlock(text = "Previous"),
                                                value = "previous",
                                                action = "previous",
                                            ),
                                            SlackButtonBlock(
                                                text = SlackTextBlock(text = "Next"),
                                                value = "next",
                                                action = "next",
                                            ),
                                        ),
                                ),
                                SlackImageBlock(
                                    url = imageUrl,
                                    alt = "GIF",
                                ),
                            ),
                        metadata = responseUrl,
                    ),
            )

        client.updateView(request)
    }
}
